src/ripple/net/HTTPStream.h

```diff
@@ -0,0 +1,158 @@
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NET_HTTPSTREAM_H_INCLUDED
+#define RIPPLE_NET_HTTPSTREAM_H_INCLUDED
+
+#include <ripple/core/Config.h>
+#include <ripple/net/HTTPClientSSLContext.h>
+
+#include <boost/asio/spawn.hpp>
+#include <boost/asio/strand.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+
+namespace ripple {
+
+class HTTPStream
+{
+public:
+    using request = boost::beast::http::request<boost::beast::http::empty_body>;
+    using parser = boost::beast::http::basic_parser<false>;
+
+    virtual ~HTTPStream() = default;
+
+    template <class T>
+    static std::unique_ptr<HTTPStream>
+    makeUnique(
+        Config const& config,
+        boost::asio::io_service::strand& strand,
+        beast::Journal j)
+    {
+        return std::make_unique<T>(config, strand, j);
+    }
+
+    [[nodiscard]] virtual boost::asio::ip::tcp::socket&
+    getStream() = 0;
+
+    [[nodiscard]] virtual bool
+    connect(
+        std::string& errorOut,
+        std::string const host,
+        std::string const port,
+        boost::asio::yield_context& yield) = 0;
+
+    virtual void
+    asyncWrite(
+        request& req,
+        boost::asio::yield_context& yield,
+        boost::system::error_code& ec) = 0;
+
+    virtual void
+    asyncRead(
+        boost::beast::flat_buffer& buf,
+        parser& p,
+        bool readSome,
+        boost::asio::yield_context& yield,
+        boost::system::error_code& ec) = 0;
+};
+
+class SSLStream : public HTTPStream
+{
+public:
+    SSLStream(
+        Config const& config,
+        boost::asio::io_service::strand& strand,
+        beast::Journal j);
+
+    virtual ~SSLStream() = default;
+
+    boost::asio::ip::tcp::socket&
+    getStream() override;
+
+    bool
+    connect(
+        std::string& errorOut,
+        std::string const host,
+        std::string const port,
+        boost::asio::yield_context& yield) override;
+
+    void
+    asyncWrite(
+        request& req,
+        boost::asio::yield_context& yield,
+        boost::system::error_code& ec) override;
+
+    void
+    asyncRead(
+        boost::beast::flat_buffer& buf,
+        parser& p,
+        bool readSome,
+        boost::asio::yield_context& yield,
+        boost::system::error_code& ec) override;
+
+private:
+    HTTPClientSSLContext ssl_ctx_;
+    boost::optional<boost::asio::ssl::stream<boost::asio::ip::tcp::socket>>
+        stream_;
+    boost::asio::io_service::strand& strand_;
+};
+
+class RawStream : public HTTPStream
+{
+public:
+    RawStream(
+        Config const& config,
+        boost::asio::io_service::strand& strand,
+        beast::Journal j);
```

Comment:
`config` and `journal` are unused. These variable should be remove from the ctor. I recommend eliminating the `makeUnique` and modifying `HTTPDownloader` as such.
eg.
```
    if (ssl)
        stream_ = std::make_unique<SSLStream>(config_, strand_, j_);
    else
        stream_ = std::make_unique<RawStream>(strand_);
```

Reply:
Done